### PR TITLE
Only log admin actions when change actually happened

### PR DESCRIPTION
**Pull request by mjankowski**
Alternative to https://github.com/mastodon/mastodon/pull/20206

First three commits are prep work (reasons should be clear from messages), last commit is actual change.

### CODE REVIEW

File: spec/controllers/admin/accounts_controller_spec.rb [hunk 1]
```diff
@@ -215,13 +215,8 @@
...
 
-        expect(latest_admin_action_log)
-          .to be_present
-          .and have_attributes(
-            action: eq(:reject),
-            account_id: eq(current_user.account_id),
-            target_id: eq(account.user.id)
-          )
+        expect { account.reload }
+          .to raise_error(ActiveRecord::RecordNotFound)
```
Comment (ClearlyClaire) [inline spec/controllers/admin/accounts_controller_spec.rb:219]: Seems like the test was missing a test on the state of the account, but is now missing the part about the log.

File: spec/requests/api/v1/admin/accounts_spec.rb [hunk 2]
```diff
@@ -199,18 +199,6 @@
...
-        subject
-
-        expect(latest_admin_action_log)
-          .to be_present
-          .and have_attributes(
-            action: eq(:reject),
-            account_id: eq(user.account_id),
-            target_id: eq(account.user.id)
-          )
-      end
```
Comment (ClearlyClaire) [inline spec/requests/api/v1/admin/accounts_spec.rb:213]: This also removes the test?

**Review by ClearlyClaire — COMMENTED** [review]
This kind of makes sense, but I'm not super confident in the change… especially with some related tests seemingly getting removed.

File: spec/controllers/admin/accounts_controller_spec.rb (hunk 1 above)
Comment (mjankowski) [reply spec/controllers/admin/accounts_controller_spec.rb:219]: Ah, thanks for asking about this, I had meant to leave inline notes.

For the two removed specs which previously checked for the action log ... these are both happening in "reject" actions, where the user record in question is sent to the `DeleteAccountService` by the controller action (with `reserve_email: false` in both cases, which means the user record will be deleted) before `log_action` is called.

With this change, the log record is no longer created ... which seems like a bugfix to me? ... hence I removed that from the spec.

If that's NOT a bug and we actively want to create a log with a reference to a record which will have been deleted (or I'm misunderstanding something here) ... let me know and I can adjust this to handle possible missing records.

I could also probably pull out 2-3 precursor PRs from this, and get this one to be JUST the changes to the concern and its spec -- everything else could be justified on its own.

Comment (mjankowski) [reply spec/controllers/admin/accounts_controller_spec.rb]: Pushed a change which restores the behavior of creating action logs for deleted records.

**Comment by mjankowski** [comment]
> This kind of makes sense, but I'm not super confident in the change… especially with some related tests seemingly getting removed.

With the specs restored and "log actions on deleted records" behavior preserved, any more concerns here?

**Comment by ClearlyClaire** [comment]
This looks ok but I'd like someone else to have a look. My concern is that an incorrect call, an abruptly aborted process, or some edge case we are not seeing, could result in a log entry missing, which is IMHO worse than having duplicate entries.

**Comment by mjankowski** [comment]
I will patiently wait on "SOMEONE ELSE" to look here.

The other option here would be to just close the issue as not valid, and claim that what we are actually capturing here is anyone's ATTEMPT to perform any of these actions, and that even if they did the same thing like 20 times in a row we actually do want to capture all of that in the logs.

This PR definitely assumes that we merely want to capture SUCCESSFUL efforts to do these actions.

**Comment by renchap** [comment]
Ideally we should rely on database transactions here, so if the transaction is cancelled then the audit log is not written. But I assume that this is complex to implement in this case?

**Comment by mjankowski** [comment]
> Ideally we should rely on database transactions here, so if the transaction is cancelled then the audit log is not written.

I'm interpreting this to mean "we only want to log successful saves which actually changed something" ... ?

> But I assume that this is complex to implement in this case?

Everything is still happening at the AR/Rails level, not DB level. We could hypothetically re-architect this to sit in after_commit hooks or something, but the knowledge of who the current user is sits at controller level and we'd need more changes to pass that in (I also think the net result might be uglier).

I have not surveyed ALL of the usage of this method, but from what I have reviewed, every spot that calls this either calls it from a "if save" branch or calls it after using `save!` (or other raising method) ... which are both already contributing to reducing the times we might audit log a failed update, and would continue to do so.

**Comment by ThisIsMissEm** [comment]
> Everything is still happening at the AR/Rails level, not DB level. We could hypothetically re-architect this to sit in after_commit hooks or something, but the knowledge of who the current user is sits at controller level and we'd need more changes to pass that in (I also think the net result might be uglier).

@mjankowski I wonder if the CurrentAttributes mechanism might be able to clean this up? https://fullstackheroes.com/tutorials/rails/current-attributes/ 

That way we'd have a global way to access the current user, including from models?

**Comment by mjankowski** [comment]
Yes, that'd be a decent approach if we wanted to move the behaviour out of controllers.

For purposes of this PR I'd like to keep it just doing the changes check.

**Comment by mjankowski** [comment]
Pulling out some things to smaller PRs here - but still need feedback on direction for this one.

**Comment by mjankowski** [comment]
This one is still in "needs a decision" pending state.

**Comment by mjankowski** [comment]
Who is the correct person to retroactively decide how this feature is supposed to work?

**Pull request closed without merging**